From 16df6a47e27d672305d1db4de20f817ab79115f3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tom=C3=A1s=20Gonz=C3=A1lez?=
 <tomasagustin.gonzalezorlando@arm.com>
Date: Thu, 14 Dec 2023 10:27:56 +0000
Subject: [PATCH] Cargo.toml: Update parsec-client and psa-crypto
 dependencies
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Update:
 * parsec-client to 0.16.0
 * psa-crypto to 0.12.0

The parsec client has changed its API, adapt to that change:

 * Save the lazy_static reference to the client as a result variable
   to adapt to the parsec client returning a result when creating it

Upstream-Status: Submitted [https://github.com/parallaxsecond/parsec-se-driver/pull/36]
Signed-off-by: Tomás González <tomasagustin.gonzalezorlando@arm.com>
---
 Cargo.toml            |  4 ++--
 src/asymmetric.rs     | 32 +++++++++++++++++++++-----------
 src/key_management.rs | 21 ++++++++++++++++-----
 src/lib.rs            | 36 +++++++++++++++++++++++-------------
 4 files changed, 62 insertions(+), 31 deletions(-)

diff --git a/Cargo.toml b/Cargo.toml
index 950d4d4..ba1ab77 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -16,9 +16,9 @@ name = "parsec_se_driver"
 crate-type = ["staticlib"]
 
 [dependencies]
-parsec-client = "0.13.0"
+parsec-client = "0.16.0"
 lazy_static = "1.4.0"
-psa-crypto = { version = "0.9.0", default-features = false, features = ["interface"] }
+psa-crypto = { version = "0.12.0", default-features = false, features = ["interface"] }
 log = "0.4.11"
 env_logger = { version = "0.7.1", optional = true }
 
diff --git a/src/asymmetric.rs b/src/asymmetric.rs
index 37f11c4..7fbdedf 100644
--- a/src/asymmetric.rs
+++ b/src/asymmetric.rs
@@ -30,11 +30,16 @@ unsafe extern "C" fn p_sign(
         Ok(alg) => alg,
         Err(e) => return e.into(),
     };
-    let signature = match PARSEC_BASIC_CLIENT.read().unwrap().psa_sign_hash(
-        &key_slot_to_key_name(key_slot),
-        std::slice::from_raw_parts(p_hash, hash_length),
-        alg,
-    ) {
+    let signature = match PARSEC_BASIC_CLIENT
+        .read()
+        .unwrap()
+        .as_ref()
+        .unwrap()
+        .psa_sign_hash(
+            &key_slot_to_key_name(key_slot),
+            std::slice::from_raw_parts(p_hash, hash_length),
+            alg,
+        ) {
         Ok(signature) => signature,
         Err(e) => return client_error_to_psa_status(e),
     };
@@ -58,12 +63,17 @@ unsafe extern "C" fn p_verify(
         Ok(alg) => alg,
         Err(e) => return e.into(),
     };
-    match PARSEC_BASIC_CLIENT.read().unwrap().psa_verify_hash(
-        &key_slot_to_key_name(key_slot),
-        std::slice::from_raw_parts(p_hash, hash_length),
-        alg,
-        std::slice::from_raw_parts(p_signature, signature_length),
-    ) {
+    match PARSEC_BASIC_CLIENT
+        .read()
+        .unwrap()
+        .as_ref()
+        .unwrap()
+        .psa_verify_hash(
+            &key_slot_to_key_name(key_slot),
+            std::slice::from_raw_parts(p_hash, hash_length),
+            alg,
+            std::slice::from_raw_parts(p_signature, signature_length),
+        ) {
         Ok(_) => PSA_SUCCESS,
         Err(e) => client_error_to_psa_status(e),
     }
diff --git a/src/key_management.rs b/src/key_management.rs
index 565888e..cdfafc5 100644
--- a/src/key_management.rs
+++ b/src/key_management.rs
@@ -53,11 +53,16 @@ unsafe extern "C" fn p_import(
         Err(e) => return e.into(),
     };
     *bits = attributes.bits;
-    match PARSEC_BASIC_CLIENT.read().unwrap().psa_import_key(
-        &key_slot_to_key_name(key_slot),
-        std::slice::from_raw_parts(data, data_length),
-        attributes,
-    ) {
+    match PARSEC_BASIC_CLIENT
+        .read()
+        .unwrap()
+        .as_ref()
+        .unwrap()
+        .psa_import_key(
+            &key_slot_to_key_name(key_slot),
+            std::slice::from_raw_parts(data, data_length),
+            attributes,
+        ) {
         Ok(_) => PSA_SUCCESS,
         Err(e) => client_error_to_psa_status(e),
     }
@@ -78,6 +83,8 @@ unsafe extern "C" fn p_generate(
     match PARSEC_BASIC_CLIENT
         .read()
         .unwrap()
+        .as_ref()
+        .unwrap()
         .psa_generate_key(&key_slot_to_key_name(key_slot), attributes)
     {
         Ok(_) => PSA_SUCCESS,
@@ -93,6 +100,8 @@ unsafe extern "C" fn p_destroy(
     match PARSEC_BASIC_CLIENT
         .read()
         .unwrap()
+        .as_ref()
+        .unwrap()
         .psa_destroy_key(&key_slot_to_key_name(key_slot))
     {
         Ok(_) => PSA_SUCCESS,
@@ -110,6 +119,8 @@ unsafe extern "C" fn p_export_public(
     let key_material = match PARSEC_BASIC_CLIENT
         .read()
         .unwrap()
+        .as_ref()
+        .unwrap()
         .psa_export_public_key(&key_slot_to_key_name(key))
     {
         Ok(key) => key,
diff --git a/src/lib.rs b/src/lib.rs
index d8777af..e4ec280 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -35,6 +35,7 @@
 mod asymmetric;
 mod key_management;
 
+use parsec_client::error::Result;
 use psa_crypto::ffi::{
     psa_drv_se_asymmetric_t, psa_drv_se_context_t, psa_drv_se_key_management_t, psa_drv_se_t,
     psa_key_location_t, psa_key_slot_number_t, psa_status_t,
@@ -57,10 +58,9 @@ use std::ptr;
 use std::sync::RwLock;
 
 lazy_static! {
-    // We do not use "new" directly here because it can fail and we do not want to panic. The
-    // automatic selection of authentication/provider is done in `p_init` so that we can return an
-    // error if there is a problem with that.
-    static ref PARSEC_BASIC_CLIENT: RwLock<BasicClient> = RwLock::new(BasicClient::new_naked());
+    // The automatic selection of authentication/provider is done in `p_init` so that we can return
+    // an error if there is a problem with that.
+    static ref PARSEC_BASIC_CLIENT: RwLock<Result<BasicClient>> = RwLock::new(BasicClient::new_naked());
 }
 
 /// Parsec SE Driver structure
@@ -86,16 +86,26 @@ unsafe extern "C" fn p_init(
     // Ignore if the initialisation failed because the `p_init` function has already been called.
     let _ = env_logger::try_init();
 
-    let mut client = (*PARSEC_BASIC_CLIENT).write().expect("lock poisoned");
+    match (*PARSEC_BASIC_CLIENT)
+        .write()
+        .expect("lock poisoned")
+        .as_mut()
+    {
+        Ok(client) => {
+            if let Err(e) = client.set_default_auth(Some(String::from("Parsec SE Driver"))) {
+                error!("Error setting the default authentication method ({}).", e);
+                return PSA_ERROR_GENERIC_ERROR;
+            }
 
-    if let Err(e) = client.set_default_auth(Some(String::from("Parsec SE Driver"))) {
-        error!("Error setting the default authentication method ({}).", e);
-        return PSA_ERROR_GENERIC_ERROR;
-    }
-
-    if let Err(e) = client.set_default_provider() {
-        error!("Error setting the default provider ({}).", e);
-        return PSA_ERROR_GENERIC_ERROR;
+            if let Err(e) = client.set_default_provider() {
+                error!("Error setting the default provider ({}).", e);
+                return PSA_ERROR_GENERIC_ERROR;
+            }
+        }
+        Err(e) => {
+            error!("Error Setting up the Parsec Client ({}).", e);
+            return PSA_ERROR_GENERIC_ERROR;
+        }
     }
 
     info!("SE Driver initialized");
-- 
2.25.1

